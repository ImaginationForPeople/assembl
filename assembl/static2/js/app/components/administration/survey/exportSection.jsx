// @flow
import React from 'react';
import { connect } from 'react-redux';
import { I18n, Translate } from 'react-redux-i18n';
import { Link } from 'react-router';
import { compose, graphql } from 'react-apollo';
import { FormGroup, Radio, FormControl } from 'react-bootstrap';

import { getDiscussionId } from '../../../utils/globalFunctions';
import SectionTitle from '../sectionTitle';
import DiscussionPreferenceLanguage from '../../../graphql/DiscussionPreferenceLanguage.graphql';

class ExportSection extends React.Component {
  static ExportLanguageDropDown = ({ languages, onSelect, activeKey }) => {
    const activeLanguage = languages.filter(language => language.locale === activeKey)[0];
    return (
      <FormControl
        className="export-language-dropdown"
        componentClass="select"
        onChange={({ target: { value } }) => {
          onSelect(value);
        }}
        value={activeLanguage.locale}
      >
        {languages.map(lang => (
          <option key={`locale-${lang.locale}`} value={lang.locale}>
            {lang.name}
          </option>
        ))}
      </FormControl>
    );
  };

  state = { exportLocale: null, translate: false };

  selectExportLocale = (locale) => {
    this.setState({ exportLocale: locale });
  };

  toggleTranslation = (shouldTranslate) => {
    this.setState({ translate: shouldTranslate });
  };

  render() {
    const { data: { discussionPreferences: { languages } } } = this.props;
    const { translate } = this.state;
    const debateId = getDiscussionId();
    if (!debateId) return null;
    const exportLocale = this.state.exportLocale || languages[0].locale;
    const exportLink = `/data/Discussion/${debateId}/phase1_csv_export${translate ? `?lang=${exportLocale}` : ''}`;
    return (
      <div className="admin-box survey-admin-export-section">
        <SectionTitle title={I18n.t('administration.survey.2')} annotation={I18n.t('administration.surveyExport.annotation')} />
        <div className="admin-content">
          <FormGroup>
            <Radio
              checked={!translate}
              onChange={() => {
                this.toggleTranslation(false);
              }}
            >
              <Translate value="administration.surveyExport.noExportLanguage" />
            </Radio>
            <Radio
              checked={translate}
              onChange={() => {
                this.toggleTranslation(true);
              }}
            >
              <Translate value="administration.surveyExport.translateTheMessagesIn" />
              {translate && (
                <ExportSection.ExportLanguageDropDown
                  languages={languages}
                  onSelect={this.selectExportLocale}
                  activeKey={exportLocale}
                />
              )}
            </Radio>
          </FormGroup>
          <br />
          <Link className="button-link button-dark margin-l" href={exportLink}>
            <Translate value="administration.surveyExport.link" />
          </Link>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ i18n }) => ({
  i18n: i18n
});

export default compose(
  connect(mapStateToProps),
  graphql(DiscussionPreferenceLanguage, {
    options: ({ i18n: { locale } }) => ({
      variables: {
        inLocale: locale
      }
    })
  })
)(ExportSection);