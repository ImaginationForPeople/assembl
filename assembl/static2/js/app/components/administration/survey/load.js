// @flow
import sortBy from 'lodash/sortBy';
import type { ApolloClient } from 'react-apollo';

import { I18n } from 'react-redux-i18n';
import ThematicsQuery from '../../../graphql/ThematicsQuery.graphql';
import { convertEntriesToI18nValue, convertEntriesToI18nRichText } from '../../form/utils';
import ThematicsDataQuery from '../../../graphql/ThematicsDataQuery.graphql';
import { type Option } from '../../form/selectFieldAdapter';
import type { ThemesAdminValues, ThemeValue, ThemeValueFromQuery } from './types.flow';
import { getTree } from '../../../utils/tree';

export const load = async (client: ApolloClient, fetchPolicy: FetchPolicy, discussionPhaseId: ?string, locale: string) => {
  const { data } = await client.query({
    query: ThematicsQuery,
    variables: { discussionPhaseId: discussionPhaseId },
    fetchPolicy: fetchPolicy
  });
  // Prefetch the ThematicsDataQuery for the admin menu
  client.query({
    query: ThematicsDataQuery,
    variables: { discussionPhaseId: discussionPhaseId, lang: locale }
  });
  return data;
};

const getMessageViewOverride = (key: ?string): Option => {
  if (key) {
    return { value: key, label: I18n.t(`administration.modules.${key}`) };
  }
  return { value: 'noModule', label: I18n.t('administration.modules.noModule') };
};

const getThemeData = (t: ThemeValueFromQuery): ThemeValue => {
  const announcement = {
    title: convertEntriesToI18nValue(t.announcement ? t.announcement.titleEntries : []),
    body: convertEntriesToI18nRichText(t.announcement ? t.announcement.bodyEntries : []),
    quote: convertEntriesToI18nRichText(t.announcement ? t.announcement.quoteEntries : []),
    summary: convertEntriesToI18nRichText(t.announcement ? t.announcement.summaryEntries : [])
  };
  const isChecked1 = t.messageColumns && t.messageColumns.length >= 1 ? t.messageColumns.length === 2 : true;
  const isChecked2 = t.messageColumns && t.messageColumns.length >= 1 ? t.messageColumns.length === 3 : false;
  return {
    id: t.id,
    messageViewOverride: getMessageViewOverride(t.messageViewOverride),
    numPosts: t.numPosts,
    img: t.img,
    questions:
      (t.questions &&
        t.questions.map(q => ({
          id: q.id,
          title: convertEntriesToI18nValue(q.titleEntries)
        }))) ||
      [],
    title: convertEntriesToI18nValue(t.titleEntries),
    description: convertEntriesToI18nValue(t.descriptionEntries),
    announcement: announcement,
    order: t.order,
    children: sortBy(t.children, 'order').map(t2 => getThemeData(t2)),
    multiColumns: {
      radioButtons: [
        { label: I18n.t('administration.tableOfThematics.multiColumnsFormName1'), isChecked: isChecked1, size: 2 },
        { label: I18n.t('administration.tableOfThematics.multiColumnsFormName2'), isChecked: isChecked2, size: 3 }
      ],
      messageColumns:
        (t.messageColumns &&
          t.messageColumns.map(col => ({
            messageClassifier: col.messageClassifier,
            title: convertEntriesToI18nValue(col.titleEntries),
            name: convertEntriesToI18nValue(col.nameEntries),
            color: col.color,
            columnSynthesis: {
              subject: col.columnSynthesis ? convertEntriesToI18nValue(col.columnSynthesis.subjectEntries) : {},
              body: col.columnSynthesis ? convertEntriesToI18nRichText(col.columnSynthesis.bodyEntries) : {}
            }
          }))) ||
        []
    }
  };
};

export function postLoadFormat(data: ThematicsQueryQuery): ThemesAdminValues {
  const { rootIdea, thematics } = data;
  // $FlowFixMe
  const tree = thematics ? getTree(rootIdea && rootIdea.id, thematics) : [];
  return {
    themes: sortBy(tree, 'order').map(t => getThemeData(t))
  };
}