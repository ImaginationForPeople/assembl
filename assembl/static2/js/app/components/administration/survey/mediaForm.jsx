import React from 'react';
import { connect } from 'react-redux';
import { Translate, I18n } from 'react-redux-i18n';
import { FormGroup, Checkbox, Button } from 'react-bootstrap';
import { graphql, compose } from 'react-apollo';

import FileUploader from '../../common/fileUploader';
import uploadDocumentMutation from '../../../graphql/mutations/uploadDocument.graphql';
import { getEntryValueForLocale } from '../../../utils/i18n';

import {
  toggleVideo,
  updateVideoHtmlCode,
  updateVideoDescriptionTop,
  updateVideoDescriptionBottom,
  updateVideoDescriptionSide,
  updateVideoTitle
} from '../../../actions/adminActions';
import FormControlWithLabel from '../../common/formControlWithLabel';

const relativeURL = uRL => uRL.match(/^https?:\/\/.*?(\/.*)$/)[1];

class MediaForm extends React.Component {
  constructor() {
    super();
    this.state = {
      mediaMimeType: ''
    };
  }

  onFileChange = (file) => {
    const { uploadDocument, updateHtmlCode } = this.props;
    uploadDocument({ variables: { file: file } }).then((res) => {
      const { externalUrl, mimeType } = res.data.uploadDocument.document;
      updateHtmlCode(relativeURL(externalUrl));
      this.setState({ mediaMimeType: mimeType });
    });
  };

  render() {
    const {
      descriptionTop,
      descriptionBottom,
      descriptionSide,
      hasMedia,
      htmlCode,
      selectedLocale,
      title,
      toggle,
      updateDescriptionTop,
      updateDescriptionBottom,
      updateDescriptionSide,
      updateTitle,
      updateHtmlCode
    } = this.props;
    const upperCaseLocale = selectedLocale.toUpperCase();
    const titlePh = `${I18n.t('administration.ph.title')} ${upperCaseLocale}`;
    const quotePh = `${I18n.t('administration.ph.quote')} ${upperCaseLocale}`;
    const descriptionTopPh = `${I18n.t('administration.ph.descriptionTop')} ${upperCaseLocale}`;
    const descriptionBottomPh = `${I18n.t('administration.ph.descriptionBottom')} ${upperCaseLocale}`;
    const mediaLinkPh = `${I18n.t('administration.ph.mediaLink')} ${upperCaseLocale}`;
    const isLocalURL = htmlCode[0] === '/';
    return (
      <div className="form-container">
        <div className="margin-xl">
          <FormGroup>
            <Checkbox checked={hasMedia} onChange={toggle}>
              <Translate value="administration.announcementModule" />
            </Checkbox>
          </FormGroup>
          {hasMedia ? (
            <div className="media-form">
              <FormControlWithLabel
                id="media-title"
                label={titlePh}
                required
                type="text"
                value={title}
                onChange={e => updateTitle(e.target.value)}
              />
              <FormControlWithLabel
                componentClass="textarea"
                id="media-description-top"
                type="rich-text"
                label={descriptionTopPh}
                value={descriptionTop}
                onChange={updateDescriptionTop}
              />
              <FormControlWithLabel
                componentClass="textarea"
                id="media-description-bottom"
                type="rich-text"
                label={descriptionBottomPh}
                value={descriptionBottom}
                onChange={updateDescriptionBottom}
              />
              <FormControlWithLabel
                componentClass="textarea"
                id="media-description-side"
                type="rich-text"
                label={quotePh}
                value={descriptionSide}
                onChange={updateDescriptionSide}
              />
              <FormControlWithLabel
                id="media-link"
                type="text"
                label={mediaLinkPh}
                value={htmlCode}
                onChange={e => updateHtmlCode(e.target.value)}
              />
              <div className="admin-help">
                <Translate value="administration.videoHelp" />
              </div>
              <Translate value="administration.ph.orAttachPicture" />
              <FileUploader
                handleChange={this.onFileChange}
                fileOrUrl={isLocalURL ? htmlCode : ''}
                mimeType={this.state.mediaMimeType}
              />
              {htmlCode &&
                isLocalURL && (
                  <div className="right">
                    <Button
                      onClick={() => {
                        fetch(htmlCode.slice(0, -5), { method: 'DELETE' }).then(() => {
                          this.props.updateHtmlCode('');
                        });
                      }}
                    >
                      <span className="assembl-icon-delete grey" />
                    </Button>
                  </div>
                )}
              <div className="separator" />
            </div>
          ) : null}
        </div>
      </div>
    );
  }
}

export const mapStateToProps = ({ admin: { thematicsById } }, { thematicId, selectedLocale }) => {
  const media = thematicsById.getIn([thematicId, 'video']);
  const hasMedia = media && media !== null;
  let descriptionTop;
  let descriptionBottom;
  let descriptionSide;
  let htmlCode = '';
  let title = '';
  if (hasMedia) {
    descriptionTop = getEntryValueForLocale(media.get('descriptionEntriesTop'), selectedLocale);
    descriptionBottom = getEntryValueForLocale(media.get('descriptionEntriesBottom'), selectedLocale);
    descriptionSide = getEntryValueForLocale(media.get('descriptionEntriesSide'), selectedLocale);
    htmlCode = media.get('htmlCode', '');
    title = getEntryValueForLocale(media.get('titleEntries'), selectedLocale, '');
  }
  return {
    descriptionTop: descriptionTop ? descriptionTop.toJS() : null,
    descriptionBottom: descriptionBottom ? descriptionBottom.toJS() : null,
    descriptionSide: descriptionSide ? descriptionSide.toJS() : null,
    hasMedia: hasMedia,
    htmlCode: htmlCode,
    title: title
  };
};

export const mapDispatchToProps = (dispatch, { selectedLocale, thematicId }) => ({
  toggle: () => dispatch(toggleVideo(thematicId)),
  updateHtmlCode: value => dispatch(updateVideoHtmlCode(thematicId, value)),
  updateDescriptionTop: value => dispatch(updateVideoDescriptionTop(thematicId, selectedLocale, value)),
  updateDescriptionBottom: value => dispatch(updateVideoDescriptionBottom(thematicId, selectedLocale, value)),
  updateDescriptionSide: value => dispatch(updateVideoDescriptionSide(thematicId, selectedLocale, value)),
  updateTitle: value => dispatch(updateVideoTitle(thematicId, selectedLocale, value))
});

export default compose(connect(mapStateToProps, mapDispatchToProps), graphql(uploadDocumentMutation, { name: 'uploadDocument' }))(
  MediaForm
);