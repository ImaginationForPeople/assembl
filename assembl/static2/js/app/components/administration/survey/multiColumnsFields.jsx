// @flow
import React from 'react';
import { Field } from 'react-final-form';
import range from 'lodash/range';
import { I18n } from 'react-redux-i18n';

import Helper from '../../common/helper';
import RadioButtonsFieldAdapter from '../../form/radioButtonsFieldAdapter';
import MultilingualTextFieldAdapter from '../../form/multilingualTextFieldAdapter';
import MultilingualRichTextFieldAdapter from '../../form/multilingualRichTextFieldAdapter';
import ColorPickerFieldAdapter from '../../form/colorPickerFieldAdapter';

type Props = {
  editLocale: string,
  fieldPrefix: string,
  nbColumnsInForm: number
};

const MultiColumnsFields = ({ editLocale, nbColumnsInForm, fieldPrefix }: Props) => (
  <React.Fragment>
    <div className="margin-l" />
    <Helper label={I18n.t('administration.tableOfThematics.columnsConfiguration')} helperUrl="/static2/img/helpers/helper_multicol.png" helperText="" classname="title" />
    <Field required name={`${fieldPrefix}.multiColumns.radioButtons`} component={RadioButtonsFieldAdapter} />
    {range(0, nbColumnsInForm).map(index => (
      <div key={`formcol-${index}`}>
        <Field
          required
          editLocale={editLocale}
          name={`${fieldPrefix}.multiColumns.messageColumns[${index}].name`}
          component={MultilingualTextFieldAdapter}
          label={`${I18n.t('administration.tableOfThematics.columnTitle')} ${index + 1}`}
        />
        <Field
          required
          editLocale={editLocale}
          name={`${fieldPrefix}.multiColumns.messageColumns[${index}].title`}
          component={MultilingualTextFieldAdapter}
          label={`${I18n.t('administration.tableOfThematics.columnName')} ${index + 1}`}
        />
        <Field
          required
          name={`${fieldPrefix}.multiColumns.messageColumns[${index}].color`}
          component={ColorPickerFieldAdapter}
          label={`${I18n.t('administration.tableOfThematics.columnColor')} ${index + 1}`}
        />
        <Field
          editLocale={editLocale}
          name={`${fieldPrefix}.multiColumns.messageColumns[${index}].columnSynthesis.subject`}
          component={MultilingualTextFieldAdapter}
          label={`${I18n.t('administration.tableOfThematics.columnSynthesisSubject')} ${index + 1}`}
        />
        <Field
          editLocale={editLocale}
          name={`${fieldPrefix}.multiColumns.messageColumns[${index}].columnSynthesis.body`}
          component={MultilingualRichTextFieldAdapter}
          label={`${I18n.t('administration.tableOfThematics.columnSynthesisBody')} ${index + 1}`}
        />
      </div>
    ))}
  </React.Fragment>
);

export default MultiColumnsFields;