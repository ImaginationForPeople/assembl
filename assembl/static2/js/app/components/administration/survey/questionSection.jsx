import React from 'react';
import { connect } from 'react-redux';
import { I18n } from 'react-redux-i18n';
import { Row, Col } from 'react-bootstrap';

import VideoForm from './videoForm';
import QuestionsForm from './questionsForm';
import SectionTitle from '../sectionTitle';

class QuestionSection extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedThematicId: props.thematics ? props.thematics[0] : ''
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!this.state.selectedThematicId) {
      this.setState({
        selectedThematicId: nextProps.thematics ? nextProps.thematics[0] : ''
      });
    }
  }

  render() {
    const { i18n, selectedLocale, thematics } = this.props;
    const selectedThematicId = this.state.selectedThematicId;
    return (
      <div className="admin-box">
        <SectionTitle i18n={i18n} phase="survey" tabId="1" annotation={I18n.t('administration.annotation')} />
        <div className="admin-content">
          <Row>
            {thematics.map((thematicId, index) => {
              const linkClassName = selectedThematicId === thematicId ? 'tab-title-active ellipsis' : 'tab-title ellipsis';
              return (
                <Col xs={12} md={Math.round(12 / thematics.length)} key={index}>
                  <a
                    className={linkClassName}
                    key={thematicId}
                    onClick={() => {
                      this.setState({ selectedThematicId: thematicId });
                    }}
                  >
                    {`${I18n.t('administration.thematic')} ${index + 1}`}
                  </a>
                </Col>
              );
            })}
          </Row>
          {selectedThematicId &&
            <Row>
              <VideoForm thematicId={selectedThematicId} selectedLocale={selectedLocale} />
              <QuestionsForm thematicId={selectedThematicId} selectedLocale={selectedLocale} />
            </Row>}
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ admin: { thematicsById, thematicsInOrder, selectedLocale }, i18n }) => {
  return {
    thematics: thematicsInOrder
      .filter((id) => {
        return !thematicsById.getIn([id, 'toDelete']);
      })
      .toArray(),
    i18n: i18n,
    selectedLocale: selectedLocale
  };
};

export default connect(mapStateToProps)(QuestionSection);