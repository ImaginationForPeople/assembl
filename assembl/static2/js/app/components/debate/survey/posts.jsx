/* eslint-disable  react/no-unused-prop-types */
// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import type { Map } from 'immutable';

import { updateContentLocale } from '../../../actions/contentLocaleActions';
import withLoadingIndicator from '../../common/withLoadingIndicator';
import FlatList from '../../common/flatList';
import Post from './post';
import PostsQuery from '../../../graphql/QuestionPostsQuery.graphql';
import { displayAlert } from '../../../utils/utilityManager';

type PostNode = {
  node: {
    id: string,
    originalLocale: string
  }
};

type PostsProps = {
  hasErrors: boolean,
  posts: {
    edges: Array<PostNode>
  },
  networkStatus: number,
  defaultContentLocaleMapping: Map,
  fetchMore: Function,
  refetch: Function,
  updateContentLocaleMapping: Function
};

export class DumbPosts extends React.Component<*, PostsProps, void> {
  props: PostsProps;

  componentWillMount() {
    this.updateContentLocaleMappingFromProps(this.props);
  }

  componentWillReceiveProps(nextProps: PostsProps) {
    if (nextProps.posts !== this.props.posts) {
      this.updateContentLocaleMappingFromProps(nextProps);
    }
  }

  updateContentLocaleMappingFromProps(props: PostsProps) {
    const { defaultContentLocaleMapping, updateContentLocaleMapping, posts } = props;
    const contentLocaleMappingData = {};
    posts.edges.forEach((edge) => {
      const post = edge.node;
      const { id, originalLocale } = post;
      const contentLocale = defaultContentLocaleMapping.get(originalLocale, originalLocale);
      contentLocaleMappingData[id] = {
        contentLocale: contentLocale,
        originalLocale: originalLocale
      };
    });

    updateContentLocaleMapping(contentLocaleMappingData);
  }

  render() {
    if (this.props.hasErrors) {
      displayAlert('danger', 'An error occured, please reload the page');
      return null;
    }
    const { networkStatus, fetchMore, refetch, posts } = this.props;
    return (
      <FlatList
        items={posts}
        ListItem={Post}
        onEndReachedThreshold={0.6}
        extractItems={entities => entities.question.posts}
        networkStatus={networkStatus}
        fetchMore={fetchMore}
        refetch={refetch}
        itemData={item => ({ id: item.node.id, originalLocale: item.node.originalLocale })}
      />
    );
  }
}

const mapStateToProps = state => ({
  defaultContentLocaleMapping: state.defaultContentLocaleMapping,
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

const mapDispatchToProps = dispatch => ({
  updateContentLocaleMapping: data => dispatch(updateContentLocale(data))
});

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(PostsQuery, {
    options: props => ({
      variables: { first: 8, after: '', id: props.questionId }
    }),
    props: ({ data }) => {
      if (data.loading) {
        return {
          loading: true,
          networkStatus: data.networkStatus
        };
      }

      if (data.error) {
        return {
          hasErrors: true,
          networkStatus: data.networkStatus
        };
      }

      const { question: { posts }, fetchMore, refetch, networkStatus } = data;

      return {
        hasErrors: false,
        posts: posts,
        fetchMore: fetchMore,
        refetch: refetch,
        networkStatus: networkStatus
      };
    }
  }),
  withLoadingIndicator({ color: 'black' })
)(DumbPosts);