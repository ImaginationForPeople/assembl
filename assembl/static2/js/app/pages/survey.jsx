// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Translate, I18n } from 'react-redux-i18n';
import { Grid } from 'react-bootstrap';
import type { Map } from 'immutable';

import { updateContentLocale } from '../actions/contentLocaleActions';
import withLoadingIndicator from '../components/common/withLoadingIndicator';
import Media from '../components/common/media';
import Header from '../components/common/header';
import Question from '../components/debate/survey/question';
import Navigation from '../components/debate/survey/navigation';
import Proposals from '../components/debate/survey/proposals';
import { getIfPhaseCompletedByIdentifier } from '../utils/timeline';
import ThematicQuery from '../graphql/ThematicQuery.graphql';
import { displayAlert } from '../utils/utilityManager';
import type { Timeline } from '../utils/timeline';
import { get as getRoute } from '../utils/routeMap';

type PostNode = {
  node: {
    id: string,
    originalLocale: string
  }
};

type QuestionType = {
  id: string,
  posts: {
    edges: Array<PostNode>
  },
  title: string
};

type SurveyProps = {
  debate: {
    debateData: {
      timeline: Timeline
    }
  },
  defaultContentLocaleMapping: Map,
  hasErrors: boolean,
  imgUrl: string,
  loading: boolean,
  media: Object, // TODO: we should add a type for media/video and use it everywhere
  questions: Array<QuestionType>,
  refetchThematic: Function,
  title: string,
  id: string,
  slug: string,
  updateContentLocaleMapping: Function
};

type SurveyState = {
  isScroll: boolean,
  questionIndex: number | null,
  showModal: boolean
};

class Survey extends React.Component<*, SurveyProps, SurveyState> {
  props: SurveyProps;

  state: SurveyState;

  constructor(props) {
    super(props);
    this.state = {
      isScroll: false,
      showModal: false,
      questionIndex: null
    };
  }

  componentWillMount() {
    this.updateContentLocaleMappingFromProps(this.props);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.questions !== this.props.questions) {
      this.updateContentLocaleMappingFromProps(nextProps);
    }
  }

  updateContentLocaleMappingFromProps(props) {
    const { defaultContentLocaleMapping, updateContentLocaleMapping, questions } = props;
    const contentLocaleMappingData = {};
    questions.forEach((question) => {
      question.posts.edges.forEach((edge) => {
        const post = edge.node;
        const { id, originalLocale } = post;
        const contentLocale = defaultContentLocaleMapping.get(originalLocale, originalLocale);
        contentLocaleMappingData[id] = {
          contentLocale: contentLocale,
          originalLocale: originalLocale
        };
      });
    });

    updateContentLocaleMapping(contentLocaleMappingData);
  }

  getIfProposals = (questions) => {
    if (!questions) return false;
    let isProposals = false;
    questions.forEach((question) => {
      if (question.posts.edges.length > 0) isProposals = true;
    });
    return isProposals;
  };

  scrollToQuestion = (isScroll, questionIndex) => {
    this.setState({
      isScroll: isScroll,
      questionIndex: questionIndex
    });
  };

  render() {
    if (this.props.hasErrors) {
      displayAlert('danger', I18n.t('error.loading'));
      return null;
    }
    const { imgUrl, media, questions, refetchThematic, title, slug } = this.props;
    const { debateData } = this.props.debate;
    const isPhaseCompleted = getIfPhaseCompletedByIdentifier(debateData.timeline, 'survey');
    const phaseUrl = `${getRoute('debate', { slug: slug, phase: 'survey' })}`;
    return (
      <div className="survey">
        <div className="relative">
          <Header title={title} imgUrl={imgUrl} identifier="survey" />
          {media && <Media {...media} />}
          <div className="questions">
            {questions &&
              questions.map((question, index) => (
                <Question
                  title={question.title}
                  index={index + 1}
                  key={index}
                  questionId={question.id}
                  scrollToQuestion={this.scrollToQuestion}
                  refetchTheme={refetchThematic}
                />
              ))}
          </div>
          {questions && (
            <Navigation
              questionsLength={questions.length}
              questionIndex={this.state.questionIndex}
              isScroll={this.state.isScroll}
              scrollToQuestion={this.scrollToQuestion}
            />
          )}
          <div className="proposals">
            <section className={isPhaseCompleted ? 'shown' : 'proposals-section'} id="proposals">
              <Grid fluid className="background-light">
                <div className="max-container">
                  <div className="question-title">
                    <div className="title-hyphen">&nbsp;</div>
                    <h1 className="dark-title-1">
                      <Translate value="debate.survey.proposalsTitle" />
                    </h1>
                  </div>
                  <div className="center">
                    {questions &&
                      questions.map((question, index) => (
                        <Proposals
                          nbPostsToShow={3}
                          title={question.title}
                          posts={question.posts.edges}
                          questionIndex={index + 1}
                          questionId={question.id}
                          phaseUrl={phaseUrl}
                          key={index}
                          refetchTheme={refetchThematic}
                        />
                      ))}
                  </div>
                  <div className="margin-xl">&nbsp;</div>
                </div>
              </Grid>
            </section>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  debate: state.debate,
  defaultContentLocaleMapping: state.defaultContentLocaleMapping,
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

const mapDispatchToProps = dispatch => ({
  updateContentLocaleMapping: data => dispatch(updateContentLocale(data))
});

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(ThematicQuery, {
    props: ({ data }) => {
      if (data.loading) {
        return {
          loading: true
        };
      }

      if (data.error) {
        return {
          hasErrors: true
        };
      }

      const { thematic: { img, questions, title, video: media }, refetch } = data;

      return {
        hasErrors: false,
        imgUrl: img.externalUrl,
        loading: false,
        media: media,
        questions: questions,
        refetchThematic: refetch,
        title: title
      };
    }
  }),
  withLoadingIndicator({ color: 'black' })
)(Survey);