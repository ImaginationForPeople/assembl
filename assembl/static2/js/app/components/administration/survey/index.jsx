// @flow
import React from 'react';
import isEqualWith from 'lodash/isEqualWith';
import { type ApolloClient, compose, withApollo } from 'react-apollo';
import { connect } from 'react-redux';
import arrayMutators from 'final-form-arrays';

import AdminForm from '../../form/adminForm';
import { compareEditorState } from '../../form/utils';
import swap from '../../form/swap';
import LoadSaveReinitializeForm from '../../form/LoadSaveReinitializeForm';
import Step1 from './step1';
import ConfigureThematicForm from './configureThematicForm';
import { load, postLoadFormat } from './load';
import { createMutationsPromises, save } from './save';
import validate from './validate';
import Loader from '../../common/loader';

type Props = {
  client: ApolloClient,
  phaseIdentifier: string,
  section: string,
  thematicId: string,
  discussionPhaseId: string,
  editLocale: string,
  locale: string
};

const loading = <Loader />;

const DumbSurveyAdminForm = ({ client, phaseIdentifier, section, thematicId, discussionPhaseId, editLocale, locale }: Props) => {
  if (!discussionPhaseId) {
    return loading;
  }
  return (
    <LoadSaveReinitializeForm
      load={(fetchPolicy: FetchPolicy) => load(client, fetchPolicy, discussionPhaseId, locale)}
      loading={loading}
      postLoadFormat={postLoadFormat}
      createMutationsPromises={createMutationsPromises(client, discussionPhaseId)}
      save={save}
      validate={validate}
      mutators={{
        ...arrayMutators,
        swap: swap
      }}
      render={({ handleSubmit, submitting, values, initialValues }) => {
        // the form is splitted to multiple subforms: We must calculate the global pristine.
        // Don't use final form pristine here
        const pristine = isEqualWith(initialValues, values, compareEditorState);
        return (
          <React.Fragment>
            <div className="admin-content">
              <AdminForm handleSubmit={handleSubmit} pristine={pristine} submitting={submitting}>
                {section === '1' && <Step1 editLocale={editLocale} locale={locale} discussionPhaseId={discussionPhaseId} />}
                {section === 'configThematics' && (
                  <ConfigureThematicForm
                    pristine={pristine}
                    phaseIdentifier={phaseIdentifier}
                    thematicId={thematicId}
                    editLocale={editLocale}
                    values={values}
                  />
                )}
              </AdminForm>
            </div>
          </React.Fragment>
        );
      }}
    />
  );
};

const mapStateToProps = state => ({
  debateId: state.context.debateId,
  editLocale: state.admin.editLocale,
  locale: state.i18n.locale
});

export default compose(
  connect(mapStateToProps),
  withApollo
)(DumbSurveyAdminForm);