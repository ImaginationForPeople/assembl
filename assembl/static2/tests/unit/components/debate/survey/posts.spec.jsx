import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { Map } from 'immutable';

import { DumbPosts } from '../../../../../js/app/components/debate/survey/posts';
import '../../../../helpers/setupTranslations';

describe('FlatList component', () => {
  it('should match Posts with data snapshot', () => {
    const props = {
      updateContentLocaleMapping: () => null,
      defaultContentLocaleMapping: Map({ en: 'en' }),
      hasErrors: false,
      posts: {
        edges: [
          {
            node: {
              title: 'Foobar'
            }
          }
        ]
      }
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbPosts {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});