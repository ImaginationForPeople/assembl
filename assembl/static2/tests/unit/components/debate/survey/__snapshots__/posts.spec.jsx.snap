// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatList component should match Posts with data snapshot 1`] = `
<FlatList
  ListItem={[Function]}
  extractItems={[Function]}
  fetchMore={undefined}
  itemData={[Function]}
  items={
    Object {
      "edges": Array [
        Object {
          "node": Object {
            "title": "Foobar",
          },
        },
      ],
    }
  }
  networkStatus={undefined}
  onEndReachedThreshold={0.6}
  refetch={undefined}
/>
`;
